vars/pushDockerImages.groovy

```diff
@@ -0,0 +1,139 @@
+  }
+  def registry = args.containsKey('registry') ? args.registry : error('pushDockerImages: registry parameter is required')
+  def secret = args.containsKey('secret') ? args.secret : error('pushDockerImages: secret parameter is required')
+  def targetNamespace = args.containsKey('targetNamespace') ? args.targetNamespace : error('pushDockerImages: targetNamespace parameter is required')
+  def version = args.containsKey('version') ? args.version : error('pushDockerImages: version parameter is required')
+  def imageName = args.containsKey('imageName') ? args.imageName : error('pushDockerImages: imageName parameter is required')
+  def arch = args.get('arch', 'amd64')
+  def variants = args.get('variants', [:])
+  def snapshot = args.get('snapshot', true)
+
+  // Transform version in a snapshot.
+  def sourceTag = version
+  def aliasVersion = ""
+  if (snapshot) {
+    // remove third number in version
+    aliasVersion = version.substring(0, version.lastIndexOf(".")) + "-SNAPSHOT"
+    sourceTag += "-SNAPSHOT"
+  }
+
+  // What docker tags are gonna be used
+  def tags = calculateTags(sourceTag, aliasVersion)
+
+  println tags
+  dockerLogin(secret: "${secret}", registry: "${registry}")
+  variants?.each { variant, sourceNamespace ->
+    tags.each { tag ->
+      // TODO:
+      // For backward compatibility let's ensure we tag only for amd64, then E2E can benefit from until
+      // they support the versioning with the architecture
+      if ("${arch}" == "amd64") {
+        doTagAndPush(registry: registry, imageName: imageName, variant: variant, sourceTag: sourceTag, targetTag: "${tag}", sourceNamespace: sourceNamespace, targetNamespace: targetNamespace)
+      }
+      doTagAndPush(registry: registry, imageName: imageName, variant: variant, sourceTag: sourceTag, targetTag: "${tag}-${arch}", sourceNamespace: sourceNamespace, targetNamespace: targetNamespace)
+    }
+  }
+}
+
+/**
+* Tag and push the source docker image. It retries to add resilience.
+*
+* @param imageName of the docker image
+* @param variant name of the variant used to build the docker image name
+* @param sourceNamespace namespace to be used as source for the docker tag command
+* @param targetNamespace namespace to be used as target for the docker tag command
+* @param sourceTag tag to be used as source for the docker tag command, usually under the 'beats' namespace
+* @param targetTag tag to be used as target for the docker tag command, usually under the 'observability-ci' namespace
+*/
+def doTagAndPush(Map args = [:]) {
+  def name = args.imageName
+  def variant = args.variant
+  def sourceTag = args.sourceTag
+  def targetTag = args.targetTag
+  def sourceNamespace = args.sourceNamespace
+  def targetNamespace = args.targetNamespace
+  def registry = args.registry
+
+  def sourceName = "${registry}/${sourceNamespace}/${name}${variant}:${sourceTag}"
+  def targetName = "${registry}/${targetNamespace}/${name}${variant}:${targetTag}"
```

Comment:
not sure is correct
```
def registry = 'docker.elastic.co'
def sourceNamespace = 'observability-ci'
def name = 'metricbeat'
variant = 'cloud'
sourceTag = '8.0.0-SNAPSHOT'

docker.elastic.co/observability-ci/metricbeatcloud:8.0.0-SNPASHOT
```

Reply:
ditto https://github.com/elastic/beats/pull/30414/files#diff-5edb5ab93fc95960129eecb2ee3d8763a9c5e02b56e5b80126e0c5a52296ee8fL315-L321

Reply:
I'd prefer hiding the `-` prefix, and let it as an implementation detail:

```groovy 
variants: [
      '' : 'beats',
      'ubi8' : 'beats',
      'cloud' : 'beats-ci',
      'complete' : 'beats',
    ],
```
and then evaluate it internally when building the image name: if the variant key is not empty, prepend the `-`. Otherwise the consumer could be adding whatever they like as variant separator, like `~cloud` or even worst `:cloud`. I know this is weird, but I'd see it better if we are controlling the inputs. We could even sanitise them, removing `:,\/;` etc...

Wdyt?

Reply:
I understand the concern, but I'd rather prefer the freedom in case there is a corner case to support something else. No strong opinion though

Reply:
Not a blocker on my side either. We have it already working in the old system and this PR is simply moving it to another shareable place 😄 I'm totally OK with keeping it as is.